api/all/src/main/java/io/opentelemetry/api/baggage/propagation/Parser.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.baggage.propagation;
+
+import io.opentelemetry.api.baggage.BaggageBuilder;
+import io.opentelemetry.api.baggage.BaggageEntryMetadata;
+
+/**
+ * Implements single-pass Baggage parsing in accordance with https://w3c.github.io/baggage/ Key /
+ * value are restricted in accordance with https://www.ietf.org/rfc/rfc2616.txt
+ *
+ * <p>Note: following aspects are not specified in RFC: - some invalid elements (key or value) -
+ * parser will include valid ones, disregard invalid - empty "value" is regarded as invalid - meta -
+ * anything besides element terminator (comma)
+ */
+class Parser {
+
+  private enum State {
+    KEY,
+    VALUE,
+    META
+  }
+
+  private final String baggageHeader;
+
+  private final Element key =
+      new Element(
+          new char[] {
+            '(', ')', '<', '>', '@', ',', ';', ':', '\\', '"', '/', '[', ']', '?', '=', '{', '}'
+          });
+  private final Element value = new Element(new char[] {'"', ',', ';', '\\'});
+  private String meta;
+
+  private State state;
+  private int metaStart;
+
+  private boolean skipToNext;
+
+  public Parser(String baggageHeader) {
+    this.baggageHeader = baggageHeader;
+    reset(0);
+  }
+
+  void parseInto(BaggageBuilder baggageBuilder) {
+    char[] chars = baggageHeader.toCharArray();
+    for (int i = 0, n = chars.length; i < n; i++) {
+      char current = chars[i];
+
+      if (skipToNext) {
+        if (current == ',') {
+          reset(i + 1);
+        }
+        continue;
+      }
+
+      switch (current) {
+        case '=':
+          {
+            if (state == State.KEY) {
+              if (key.tryTerminating(i, baggageHeader)) {
+                setState(State.VALUE, i + 1);
+              } else {
+                skipToNext = true;
+              }
+            }
+            break;
+          }
+        case ';':
+          {
+            if (state == State.VALUE) {
+              skipToNext = !value.tryTerminating(i, baggageHeader);
+              setState(State.META, i + 1);
+            }
+            break;
+          }
+        case ',':
+          {
+            switch (state) {
+              case VALUE:
+                value.tryTerminating(i, baggageHeader);
+                break;
+              case META:
+                meta = baggageHeader.substring(metaStart, i).trim();
```

Comment:
Not a huge deal but is it possible to avoid the trim? Can be a followup anyways

Reply:
I'd leave that as a followup. According to the spec (https://w3c.github.io/baggage/#baggage-string) OWS has to be trimmed from the meta upon parsing. Therefore to avoid explicit trim() the parser would need to keep meta in a separate, tracked data structure (just as it does for key and value).